sklearn/feature_extraction/text.py

```diff
@@ -913,6 +916,170 @@ def _make_int_array():
+        self.smooth_idf = smooth_idf
+        self.k = float(k)
+        self.b = float(b)
+
+    def fit(self, X, y=None):
+        """Learn the idf (n_samples) and beta vectors (n_features)
+
+        Parameters
+        ----------
+        X : sparse matrix, [n_samples, n_features]
+            a matrix of term/token counts
+        """
+
+        check_array(X, accept_sparse=['csr', 'csc'])
+
+        if not sp.issparse(X):
+            X = sp.csc_matrix(X)
+
+        n_samples, n_features = X.shape
+
+        # if X is an array of zeros, raise ValueError
+        if X.sum() == 0:
+            raise ValueError("X is an array of zeros")
+
+        # raise value error if there are negative values in X
+        check_non_negative(X, "Bm25Transformer")
+
+        df = _document_frequency(X)
+
+        # perform idf smoothing if required
+        df += int(self.smooth_idf)
+        n_samples_calc = n_samples + int(self.smooth_idf)
+
+        # log+1 instead of log makes sure terms with zero idf don't get
+        # suppressed entirely.
+        idf = np.log(float(n_samples_calc) / df) + 1.0
+        self._idf_diag = sp.spdiags(idf, diags=0, m=n_features, n=n_features)
+
+        # sum along rows for document lengths
+        lengths = np.array(X.sum(axis=1)).reshape((n_samples,))
+        avglen = sum(lengths) / n_samples
+
+        beta = (1 - self.b + self.b * lengths / avglen)
+        self._beta_diag = sp.spdiags(beta, diags=0, m=n_samples, n=n_samples)
+
+        return self
+
+    def transform(self, X, copy=True):
+        """Transform a count matrix to a bm25 representation
+
+        Parameters
+        ----------
+        X : sparse matrix, [n_samples, n_features]
+            a matrix of term/token counts
+
+        copy : boolean, default True
+            Whether to copy X and operate on the copy or perform in-place
+            operations.
+
+        Returns
+        -------
+        vectors : sparse matrix, [n_samples, n_features]
+        """
+
+        if hasattr(X, 'dtype') and np.issubdtype(X.dtype, np.float):
+            # preserve float family dtype
+            X = sp.csr_matrix(X, copy=copy)
+        else:
+            # convert counts or binary occurrences to floats
+            X = sp.csr_matrix(X, dtype=np.float64, copy=copy)
+
+        n_samples, n_features = X.shape
+
+        check_is_fitted(self, '_idf_diag', 'idf vector is not fitted')
+        check_is_fitted(self, '_beta_diag', 'beta vector is not fitted')
+
+        expected_n_features = self._idf_diag.shape[0]
+
+        if n_features != expected_n_features:
+            raise ValueError("Input has n_features=%d while the model"
+                             " has been trained with n_features=%d" % (
+                                 n_features, expected_n_features))
+
+        weightedtfs = X.copy()
+
+        binary = X>0
+
+        weightedtfs.data = (self.k + 1) / \
+                           (self.k * self._beta_diag.dot(binary).data / X.data + 1)
```

Comment:
self._beta_diag is created from sp.spdiags() method. I see what you mean about the API not guaranteeing the type of diagonal matrix being created.

Looking at the API docs for that method:
http://docs.scipy.org/doc/scipy/reference/generated/scipy.sparse.spdiags.html

Note it says:

> > format : str, optional
> Format of the result. By default (format=None) an appropriate sparse matrix format is returned. This choice is subject to change.

I suspect that a different sparse type is being chosen as a return value in Python 2.7 and 2.6 vs Python 3.5. I added some print statements to check as a quick and dirty way to work through it. But I'm also wondering if the X>0 is breaking something as well. 

I might need to add format="csr"